libraries/standard/mqtt/src/mqtt.c

```diff
@@ -890,6 +1211,196 @@ static MQTTStatus_t receiveConnack( MQTTContext_t * const pContext,
 
 /*-----------------------------------------------------------*/
 
+static MQTTStatus_t resendPendingAcks( MQTTContext_t * const pContext )
+{
+    MQTTStatus_t status = MQTTSuccess;
+
+    assert( pContext != NULL );
+
+    /* Check if there is any pending acks that need to be resend.
+     * Library owns resending PUBREL for a QoS2 outgoing publish.
+     * PUBREL packet will be resend if there are any packets at state
+     * #MQTTPubCompPending or #MQTTPubRelSend. */
+    status = resendPubrelForPacketInState( pContext,
+                                           MQTTPubCompPending );
+
+    if( status == MQTTSuccess )
+    {
+        /* State will be advanced to #MQTTPubCompPending after
+         * successfully sending PUBREL ack. */
+        status = resendPubrelForPacketInState( pContext,
+                                               MQTTPubRelSend );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t resendPubrelForPacketInState( MQTTContext_t * const pContext,
+                                                  MQTTPublishState_t state )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTStateCursor_t cursor = MQTT_STATE_CURSOR_INITIALIZER;
+    uint16_t packetId = MQTT_PACKET_ID_INVALID;
+    bool stateNeedUpdate = false;
+
+    assert( pContext != NULL );
+    assert( state == MQTTPubCompPending || state == MQTTPubRelSend );
+
+    /* Resending PUBREL packet. */
+    MQTTPublishState_t publishRecordState = MQTTPubRelSend;
+
+    /* State machine needs update only if packet is not already in state
+     * #MQTTPubCompPending. */
+    if( state == MQTTPubRelSend )
+    {
+        stateNeedUpdate = true;
+    }
+
+    do
+    {
+        /* Find the packet ids which are in given state.
+         * Using the same cursor will help restart the search from the
+         * last searched position. */
+        packetId = MQTT_StateSelect( pContext, state, &cursor );
+
+        /* Check if a valid packet id is obtained. */
+        if( packetId != MQTT_PACKET_ID_INVALID )
+        {
+            /* Resend pending PUBREL. */
+            status = sendPublishAcks( pContext,
+                                      packetId,
+                                      &publishRecordState,
```

Comment:
Do we expect `sendPublishAcks` to update `publishRecordState`? In that case, shouldn't it be reset to `MQTTPubRelSend` in every iteration of the loop?

Reply:
Yes.good catch